Collect tokens into a `HashMap` instead of a `Vec` in the language server.

JoshuaBatty opened the pull request:
This introduces the following type for collecting tokens during parsing. 

```
pub type TokenMap = HashMap<(Ident, Span), TokenType>;

#[derive(Debug, Clone)]
pub struct TokenType {
    pub parsed: AstToken,
    pub typed: Option<TypedAstToken>,
}
```

First, we traverse the untyped parse tree and collect the `AstToken` into `parsed`.  Then we traverse the typed_parse_tree and optionally collect a `TypedAstToken` if one was found. After both traversal stages, we end up with a `HashMap` that contains all of the tokens found in the sway file. 

This PR replaces the current `Token` module with the above data structure. As such, quite a lot of logic around the current capabilites needs to be re-written to accomodate for the new `TokenMap ` type. Marking this as a draft for now but opening a PR for visiblity.

adlerjohn reviewed (commented):
Off the top of my head, we've had issues with using hash maps before due to their non-determinism. See #459

mitchmindtree commented:
> Off the top of my head, we've had issues with using hash maps before due to their non-determinism

True, worth keeping in mind @JoshuaBatty that `BTreeMap` is a safer default in this regard. The non-determinism of `HashMap` comes into play when iterating over it, whereas `BTreeMap`'s keys are always ordered. If we don't iterate over the `HashMap` at all no need to worry, but worth keeping in mind.

JoshuaBatty commented:
Ok good to know, I did originally set out to use a `BTreeMap` but i'm unable to use `Span` as the key.

`the trait bound `sway_types::Span: Ord` is not satisfied`

mitchmindtree commented on sway-lsp/src/capabilities/document_symbol.rs:
```diff
@@ -21,30 +32,113 @@ pub fn to_symbol_information(tokens: &[Token], url: Url) -> Vec<SymbolInformatio
+                TypedDeclaration::ImplTrait { .. } => SymbolKind::INTERFACE,
+                TypedDeclaration::AbiDeclaration(_) => SymbolKind::INTERFACE,
+                TypedDeclaration::GenericTypeForFunctionScope { .. } => SymbolKind::TYPE_PARAMETER,
+                // currently we return `variable` type as default
```
Weird, is there no keyword symbol kind or anything long those lines? Or should we possibly be returning `None` or similar? Not a blocker as I can see this was the original behaviour, just curious.

JoshuaBatty replied:
Na there isn't a keyword variant. I agree though that we should scrutinize exactly the role and mappings of nodes to SymbolKind's in a future PR. I was just keeping the original behavior in place for now.

mitchmindtree commented on sway-lsp/src/capabilities/document_symbol.rs:
```diff
@@ -21,30 +32,113 @@ pub fn to_symbol_information(tokens: &[Token], url: Url) -> Vec<SymbolInformatio
 #[allow(warnings)]
 // TODO: the "deprecated: None" field is deprecated according to this library
-fn create_symbol_info(token: &Token, url: Url) -> SymbolInformation {
+fn create_symbol_info(ident: &Ident, token: &TokenType, url: Url) -> SymbolInformation {
```
```suggestion
fn symbol_info(ident: &Ident, token: &TokenType, url: Url) -> SymbolInformation {
```

JoshuaBatty replied:
Nice good call, i've gone ahead and applied this change and the subsequent suggestions of removing the get_ and create_ prefexies.

mitchmindtree commented on sway-lsp/src/capabilities/hover.rs:
```diff
@@ -1,98 +1,108 @@
-                VarBody::Type(var_type) => var_type.clone(),
-                _ => "".into(),
-            };
+fn get_hover_format(token: &TokenType, ident: &Ident) -> Hover {
```
```suggestion
fn hover_format(token: &TokenType, ident: &Ident) -> Hover {
```

mitchmindtree commented on sway-lsp/src/capabilities/hover.rs:
```diff
@@ -1,98 +1,108 @@
+    };
+
+    let format_variable_hover = |is_mutable: bool, type_name: String| -> String {
+        format!("let{} {}: {}", is_mutable, token_name, type_name,)
```
It looks like this would format as something like:

```rust
letfalse token_name: type_name
```

Is that the intent?

JoshuaBatty replied:
Nope that's the intent :) good spot!

mitchmindtree commented on sway-lsp/src/capabilities/semantic_tokens.rs:
```diff
@@ -28,43 +34,51 @@ pub fn get_semantic_tokens_full(
 }
 
-fn create_semantic_token(next_token: &Token, prev_token: Option<&Token>) -> SemanticToken {
+fn create_semantic_token(
```
```suggestion
fn semantic_token(
```

mitchmindtree commented on sway-lsp/src/capabilities/semantic_tokens.rs:
```diff
@@ -28,43 +34,51 @@ pub fn get_semantic_tokens_full(
     let token_modifiers_bitset = 0;
-    let token_type = get_type(&next_token.token_type);
-    let length = next_token.length;
+    let length = next_token_range.end.character - next_token_range.start.character;
```
```suggestion
    let length = next_token_range.len_chars();
```
^ Maybe this range type should have a `len_chars()` method or something along these lines?

JoshuaBatty replied:
Yeah that would be cool. The `next_token_range` here is actually a `Range` type in the `lsp_types` crate though.

mitchmindtree commented on sway-lsp/src/capabilities/semantic_tokens.rs:
```diff
@@ -75,27 +89,45 @@ fn create_semantic_token(next_token: &Token, prev_token: Option<&Token>) -> Sema
-        TokenType::FunctionParameter => TokenTypeIndex::Parameter as u32,
-        TokenType::VariableDeclaration(_) | TokenType::VariableExpression => {
-            TokenTypeIndex::Variable as u32
+fn get_type_idx(ast_token: &AstToken) -> u32 {
```
```suggestion
fn type_idx(ast_token: &AstToken) -> u32 {
```

mitchmindtree commented on sway-lsp/src/core/document.rs:
```diff
@@ -38,84 +34,93 @@ impl TextDocument {
-                    return Some(token);
-                }
+    /// Check if the code editor's cursor is currently over an of our collected tokens
+    pub fn get_token_at_position(&self, position: Position) -> Option<(Ident, &TokenType)> {
```
```suggestion
    pub fn token_at_position(&self, position: Position) -> Option<(Ident, &TokenType)> {
```

mitchmindtree commented on sway-lsp/src/core/document.rs:
```diff
@@ -38,84 +34,93 @@ impl TextDocument {
-        } else {
-            None
-        }
+    pub fn get_all_references(&self, token: &TokenType) -> Vec<(&Ident, &TokenType)> {
```
```suggestion
    pub fn all_references(&self, token: &TokenType) -> Vec<(&Ident, &TokenType)> {
```

mitchmindtree commented on sway-lsp/src/core/document.rs:
```diff
@@ -38,84 +34,93 @@ impl TextDocument {
-                let token = &self.tokens[*index];
-                if token.is_initial_declaration() {
-                    return Some(token);
+    pub fn get_declared_token_ident(&self, token: &TokenType) -> Option<Ident> {
```
```suggestion
    pub fn declared_token_ident(&self, token: &TokenType) -> Option<Ident> {
```

mitchmindtree commented on sway-lsp/src/core/document.rs:
```diff
@@ -38,84 +34,93 @@ impl TextDocument {
     }
 
-    pub fn _get_token_map(&self) -> &TokenMap {
+    pub fn get_token_map(&self) -> &TokenMap {
```
```suggestion
    pub fn token_map(&self) -> &TokenMap {
```

mitchmindtree commented on sway-lsp/src/core/session.rs:
```diff
@@ -74,15 +76,14 @@ impl Session {
     // Token
     pub fn get_token_ranges(&self, url: &Url, position: Position) -> Option<Vec<Range>> {
```
```suggestion
    pub fn token_ranges(&self, url: &Url, position: Position) -> Option<Vec<Range>> {
```

mitchmindtree commented on sway-lsp/src/core/traverse_parse_tree.rs:
```diff
@@ -0,0 +1,420 @@
+}
+
+fn handle_custom_type(_type_info: &TypeInfo, _tokens: &mut TokenMap) {
+    // TODO: Not obvious how to handle this now with the new types
```
```suggestion
    todo!("Not obvious how to handle this now with the new types");
```

If we don't want to panic here, we should maybe open an issue for this and mention it under this `TODO` and move the commented out code to the issue?

JoshuaBatty replied:
Good idea, I've opened #2274 and moved the code over in there for the time being.

mitchmindtree commented on sway-lsp/src/utils/token.rs:
```diff
@@ -0,0 +1,64 @@
+    (ident.clone(), ident.span())
+}
+
+pub fn get_type_id(token_type: &TokenType) -> Option<TypeId> {
```
```suggestion
pub fn type_id(token_type: &TokenType) -> Option<TypeId> {
```

mitchmindtree reviewed (commented):
Looking good for the most part! Just a few small comments and clarifications.

I also added a bunch of name change recommendations (mostly removing extraneous `get_` prefixes) but feel free to dismiss these comments in favour of a future PR if you'd prefer.

JoshuaBatty commented:
> Looking good for the most part! Just a few small comments and clarifications.
> 
> I also added a bunch of name change recommendations (mostly removing extraneous `get_` prefixes) but feel free to dismiss these comments in favour of a future PR if you'd prefer.

Thanks for the review @mitchmindtree! Just addressed all your feedback :)

sezna commented:
+1 on hashmap nondeterminism, but in stuff outside of core compiler functionality, the nondeterminism should be less catastrophic

mitchmindtree reviewed (approved):
:shipit:
